src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+        // fill up e.g. If a line filled up half the buffer we would not know about
+        // it until the buffer is returned. We could increase the read buffer if we are
+        // willing to accept some blocking waiting for it to fill up. I'm not sure if
+        // the underlying BufReader will read in larger chunks or not.
+        // Possibly could look into https://crates.io/crates/fixed-buffer-tokio but
+        // I don't know much about this library.
+        let mut read_buffer = [0_u8; 1];
+
+        // Not using the shutdown signal in this method so we need to shutdown
+        // in other methods to end the reading
+        while let Ok(bytes_read) = reader.read(&mut read_buffer).await {
+            if bytes_read == 0 {
+                info!("End of input reached, stop reading.");
+                break;
+            } else {
+                let read_byte = read_buffer[0];
+
+                // Could be enhanced to split 'lines' based on user defined
+                // delimiters in addition to newline.
+                if event_per_line && read_byte == b'\n' {
+                    if buffer.ends_with(&[b'\r']) {
+                        let _ = buffer.pop();
+                    }
+
+                    if let Some(buffer_string) = buffer_to_string(&mut buffer, false) {
+                        if sender.send((buffer_string, stream)).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        info!("Invalid utf8, stop reading.");
+                        break;
+                    }
+                } else {
+                    buffer.push(read_byte);
+
+                    if buffer.len() == buf_size {
+                        if let Some(buffer_string) = buffer_to_string(&mut buffer, true) {
+                            if sender.send((buffer_string, stream)).await.is_err() {
+                                break;
+                            }
+                        } else {
+                            info!("Invalid utf8, stop reading.");
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        // Handle any left over buffer
+        if !buffer.is_empty() {
+            if let Some(buffer_string) = buffer_to_string(&mut buffer, true) {
+                let _ = sender.send((buffer_string, stream)).await;
+                if !buffer.is_empty() {
+                    info!("Invalid utf8, left in buffer.");
+                }
+            } else {
+                info!("Invalid utf8, left in buffer.");
+            }
+        }
+
+        info!("Finished capturing {} command output.", stream);
+    }));
+}
+
+fn buffer_to_string(buffer: &mut Vec<u8>, allow_shrinking: bool) -> Option<String> {
+    let mut left_over_buffer: Vec<u8> = Vec::new();
+    loop {
+        if let Ok(buffer_string) = String::from_utf8(buffer.clone()) {
```

Comment:
`String::from_utf8` returns an error that can provide an indication of where the UTF-8 conversion failed (see [`Utf8Error::valid_up_to`](https://doc.rust-lang.org/std/str/struct.Utf8Error.html#method.valid_up_to)). I think this function could be rewritten with that in mind to avoid most of the overhead here.

Reply:
Thank you for pointing that out, I think that would be a really elegant way to implement this function. Per a suggestion from @jszwedko I have removed the UTF-8 logic for now. If you and @jszwedko would like be to add an `encoding` config option (e.g. binary or utf8 could be the initial options) I would be happy to add this in (and use your suggested `valid_up_to`).